### PR TITLE
Clarify the overall result in the GitHub comment.

### PR DESCRIPTION
**Pull request by nnethercote**
Specifics:
- Adds a big "Overall result" heading at the top of the comment, which
  includes "ACTION NEEDED" when there are regressions.
- Moves the next steps and "if you disagree..." text to the top of the
  comment, just below the overall result, before the tables of
  measurements. This means all the instructions that people need to read
  are in one place.
- Removes the short summaries in front of each table, because (a) they just
  duplicates what the table says, and (b) saying "regressions found" for
  cycles and max-rss is a bit misleading because it doesn't contribute
  to the overall conclusion. Replaces them with a sentence indicating if
  the measurements contributed to the overall result and why.
- Changes `Direction::join` to a non-static method, it's nicer that way.

### CODE REVIEW

**Comment by nnethercote** [comment]
Here is a before and after comparison of a real comment.

----

Finished benchmarking commit (b998821e4c51c44a9ebee395c91323c374236bbb): [comparison url](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=instructions:u)

- Primary benchmarks: ❌ relevant regressions found
- Secondary benchmarks: ❌ relevant regression found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions ❌ <br /> (primary) | 0.3%     | 0.9% | 6         |
| Regressions ❌ <br /> (secondary) | 0.2%     | 0.2% | 1         |
| Improvements ✅ <br /> (primary) | -        | -   | 0         |
| Improvements ✅ <br /> (secondary) | -        | -   | 0         |
| All ❌✅ (primary) | 0.3%     | 0.9% | 6         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: ❌ relevant regression found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions ❌ <br /> (primary) | 4.3%     | 4.3% | 1         |
| Regressions ❌ <br /> (secondary) | 7.3%     | 7.3% | 1         |
| Improvements ✅ <br /> (primary) | -        | -   | 0         |
| Improvements ✅ <br /> (secondary) | -3.7%    | -4.2% | 3         |
| All ❌✅ (primary) | 4.3%     | 4.3% | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance

----

Finished benchmarking commit (b998821e4c51c44a9ebee395c91323c374236bbb): [comparison URL](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED

**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

### [Instruction count](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=instructions:u)

This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions ❌ <br /> (primary) | 0.3%     | 0.9% | 6         |
| Regressions ❌ <br /> (secondary) | 0.2%     | 0.2% | 1         |
| Improvements ✅ <br /> (primary) | -        | -   | 0         |
| Improvements ✅ <br /> (secondary) | -        | -   | 0         |
| All ❌✅ (primary) | 0.3%     | 0.9% | 6         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=2ed0f29168f5bc116e747152be600519b5d78bfd&end=b998821e4c51c44a9ebee395c91323c374236bbb&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions ❌ <br /> (primary) | 4.3%     | 4.3% | 1         |
| Regressions ❌ <br /> (secondary) | 7.3%     | 7.3% | 1         |
| Improvements ✅ <br /> (primary) | -        | -   | 0         |
| Improvements ✅ <br /> (secondary) | -3.7%    | -4.2% | 3         |
| All ❌✅ (primary) | 4.3%     | 4.3% | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

File: site/src/comparison.rs [hunk 1]
```diff
@@ -1248,15 +1238,24 @@ pub enum Direction {
...
+        match (self, other) {
             (Self::None, b) => b,
             (a, Self::None) => a,
             (Self::Improvement, Self::Improvement) => Self::Improvement,
             (Self::Regression, Self::Regression) => Self::Regression,
             _ => Self::Mixed,
         }
     }
+
+    pub fn overall_result(&self) -> &'static str {
```
Comment (Kobzol) [inline site/src/comparison.rs:1251]: nit: This function contains text specific to the GH comments, so I think that it should be a function in `comparison_summary`, not a method on `Direction`.

File: site/src/github/comparison_summary.rs [hunk 2]
```diff
@@ -197,52 +218,25 @@ fn write_metric_summary(
             .push_str("This benchmark run did not return any relevant results for this metric.\n");
         false
     } else {
-        let primary_short_summary = generate_short_summary(&primary);
-        let secondary_short_summary = generate_short_summary(&secondary);
-
-        if hidden {
+        if !hidden {
```
Comment (Kobzol) [inline site/src/github/comparison_summary.rs:221]: nit: Maybe `hidden` should be renamed, because now the distinction is basically reliable vs unreliable?

File: site/src/github/comparison_summary.rs [hunk 3]
```diff
@@ -286,32 +281,6 @@ for rollup, we strongly recommend not doing so since this PR may lead to changes
 compiler perf.{next_steps}
 
 @bors rollup=never
-@rustbot label: +S-waiting-on-review -S-waiting-on-perf {label}",
+@rustbot label: +S-waiting-on-review -S-waiting-on-perf {sign}perf-regression",
```
Comment (Kobzol) [inline site/src/github/comparison_summary.rs:284]: This is not relevant for this PR, but I wonder if adding `+S-waiting-on-review` unconditionally is a good idea, sometimes the PR is still being developed without needing a review and the author wants to run a perf. run.

**Review by Kobzol — APPROVED** [review]
Looks great! I like that the actions to do and especially the overall is are now clearly visible at the beginning.

File: site/src/github/comparison_summary.rs [hunk 4]
```diff
@@ -133,10 +133,40 @@ async fn summarize_run(
...
+    let next_steps = if is_master_commit {
+        master_run_body(is_regression)
+    } else {
+        try_run_body(is_regression)
+    };
+    writeln!(&mut message, "{next_steps}\n").unwrap();
+
+    const DISAGREEMENT: &str =
+        "If you disagree with this performance assessment, please file an issue in \
+        [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).";
```
Comment (rylev) [inline site/src/github/comparison_summary.rs:167]: This was originally written when we moved to a new system for classifying regressions. As we've become more confident that the regression classification is not totally off base, I wonder if we should rephrase this to something more like "If you have questions about this performance assessment, please file an issue..."

Comment (nnethercote) [reply site/src/github/comparison_summary.rs:167]: Or I could just remove it. Has anyone ever filed such an issue?

Comment (rylev) [reply site/src/github/comparison_summary.rs:167]: Yes, though it's hard to tell if they filed an issue specifically because of this prompt or because they knew were to go already.

File: site/src/github/comparison_summary.rs (hunk 3 above)
Comment (nnethercote) [reply site/src/github/comparison_summary.rs:290]: Hmm, not sure. Sometimes it's the right thing to do, sometimes not.

**Comment by nnethercote** [comment]
I have addressed the review comments. I also added " - no action needed" to the heading in the case where no action is needed, for a clearer contrast to the " - ACTION NEEDED" case.

**Review by Kobzol — APPROVED** [review]
